File: scripts/middle-math/zip_converter.py
#!/usr/bin/env python3
"""Utilities for converting between emoji and numeric PPL± zip codes."""

from __future__ import annotations

import argparse
from typing import Dict, Tuple

ORDERS: Dict[int, Tuple[str, str, str]] = {
    1: ("🐂", "Foundation", "foundation"),
    2: ("⛽", "Strength", "strength"),
    3: ("🦋", "Hypertrophy", "hypertrophy"),
    4: ("🏟", "Performance", "performance"),
    5: ("🌾", "Full Body", "full-body"),
    6: ("⚖", "Balance", "balance"),
    7: ("🖼", "Restoration", "restoration"),
}

AXES: Dict[int, Tuple[str, str, str]] = {
    1: ("🏛", "Basics", "basics"),
    2: ("🔨", "Functional", "functional"),
    3: ("🌹", "Aesthetic", "aesthetic"),
    4: ("🪐", "Challenge", "challenge"),
    5: ("⌛", "Time", "time"),
    6: ("🐬", "Partner", "partner"),
}

TYPES: Dict[int, Tuple[str, str, str]] = {
    1: ("🛒", "Push", "push"),
    2: ("🪡", "Pull", "pull"),
    3: ("🍗", "Legs", "legs"),
    4: ("➕", "Plus", "plus"),
    5: ("➖", "Ultra", "ultra"),
}

COLORS: Dict[int, Tuple[str, str, str]] = {
    1: ("⚫", "Teaching", "teaching"),
    2: ("🟢", "Bodyweight", "bodyweight"),
    3: ("🔵", "Structured", "structured"),
    4: ("🟣", "Technical", "technical"),
    5: ("🔴", "Intense", "intense"),
    6: ("🟠", "Circuit", "circuit"),
    7: ("🟡", "Fun", "fun"),
    8: ("⚪", "Mindful", "mindful"),
}

PREPARATORY_COLORS = {"⚫", "🟢", "⚪", "🟡"}
EXPRESSIVE_COLORS = {"🔵", "🟣", "🔴", "🟠"}

OPERATOR_TABLE: Dict[str, Tuple[Tuple[str, str], Tuple[str, str]]] = {
    "🏛": (("📍", "pono"), ("🤌", "facio")),
    "🔨": (("🧸", "fero"), ("🥨", "tendo")),
    "🌹": (("👀", "specio"), ("🪢", "plico")),
    "🪐": (("🪵", "teneo"), ("🏹", "mitto")),
    "⌛": (("🎻", "duco"), ("✍️", "grapho")),
    "🐬": (("🫴", "capio"), ("🧿", "logos")),
}


def _emoji_to_position(table: Dict[int, Tuple[str, str, str]]) -> Dict[str, int]:
    return {emoji: position for position, (emoji, _, _) in table.items()}


ORDER_BY_EMOJI = _emoji_to_position(ORDERS)
AXIS_BY_EMOJI = _emoji_to_position(AXES)
TYPE_BY_EMOJI = _emoji_to_position(TYPES)
COLOR_BY_EMOJI = _emoji_to_position(COLORS)


def emoji_to_numeric(emoji_zip: str) -> str:
    chars = list(emoji_zip)
    if len(chars) != 4:
        raise ValueError("emoji zip must contain exactly 4 emojis")

    try:
        order = ORDER_BY_EMOJI[chars[0]]
    except KeyError as exc:
        raise ValueError(f"invalid order emoji: {chars[0]}") from exc
    try:
        axis = AXIS_BY_EMOJI[chars[1]]
    except KeyError as exc:
        raise ValueError(f"invalid axis emoji: {chars[1]}") from exc
    try:
        type_position = TYPE_BY_EMOJI[chars[2]]
    except KeyError as exc:
        raise ValueError(f"invalid type emoji: {chars[2]}") from exc
    try:
        color = COLOR_BY_EMOJI[chars[3]]
    except KeyError as exc:
        raise ValueError(f"invalid color emoji: {chars[3]}") from exc

    return f"{order}{axis}{type_position}{color}"


def _validate_numeric_or_raise(numeric_zip: str) -> None:
    if len(numeric_zip) != 4 or not numeric_zip.isdigit():
        raise ValueError("numeric zip must be exactly 4 digits")
    ranges = ((1, 7), (1, 6), (1, 5), (1, 8))
    labels = ("order", "axis", "type", "color")
    for idx, (label, (low, high)) in enumerate(zip(labels, ranges)):
        digit = int(numeric_zip[idx])
        if not low <= digit <= high:
            raise ValueError(f"{label} {digit} out of range {low}-{high}")


def numeric_to_emoji(numeric_zip: str) -> str:
    _validate_numeric_or_raise(numeric_zip)
    order, axis, type_position, color = (int(ch) for ch in numeric_zip)
    return (
        f"{ORDERS[order][0]}{AXES[axis][0]}"
        f"{TYPES[type_position][0]}{COLORS[color][0]}"
    )


def validate_zip(zip_str: str) -> bool:
    try:
        if zip_str.isdigit():
            _validate_numeric_or_raise(zip_str)
        else:
            emoji_to_numeric(zip_str)
    except ValueError:
        return False
    return True


def zip_to_deck(zip_str: str) -> int:
    numeric = zip_str if zip_str.isdigit() else emoji_to_numeric(zip_str)
    _validate_numeric_or_raise(numeric)
    order = int(numeric[0])
    axis = int(numeric[1])
    return (order - 1) * 6 + axis


def zip_to_path(emoji_zip: str) -> str:
    numeric = emoji_to_numeric(emoji_zip)
    order, axis, type_position = (int(numeric[0]), int(numeric[1]), int(numeric[2]))
    order_emoji, _, order_slug = ORDERS[order]
    axis_emoji, _, axis_slug = AXES[axis]
    type_emoji, _, type_slug = TYPES[type_position]
    return f"cards/{order_emoji}-{order_slug}/{axis_emoji}-{axis_slug}/{type_emoji}-{type_slug}/"


def derive_operator(emoji_zip: str) -> tuple[str, str]:
    chars = list(emoji_zip)
    if len(chars) != 4:
        raise ValueError("emoji zip must contain exactly 4 emojis")
    axis_emoji = chars[1]
    color_emoji = chars[3]

    if axis_emoji not in OPERATOR_TABLE:
        raise ValueError(f"invalid axis emoji: {axis_emoji}")
    if color_emoji not in COLOR_BY_EMOJI:
        raise ValueError(f"invalid color emoji: {color_emoji}")

    preparatory, expressive = OPERATOR_TABLE[axis_emoji]
    if color_emoji in PREPARATORY_COLORS:
        return preparatory
    if color_emoji in EXPRESSIVE_COLORS:
        return expressive
    raise ValueError(f"unable to derive polarity for color: {color_emoji}")


def _invalid_reason(zip_str: str) -> str:
    try:
        if zip_str.isdigit():
            _validate_numeric_or_raise(zip_str)
        else:
            emoji_to_numeric(zip_str)
    except ValueError as exc:
        return str(exc)
    return ""


def _run_convert(value: str) -> None:
    if value.isdigit():
        numeric = value
        emoji = numeric_to_emoji(numeric)
        print(f"emoji={emoji}", end=", ")
    else:
        emoji = value
        numeric = emoji_to_numeric(emoji)
        print(f"numeric={numeric}", end=", ")

    deck = zip_to_deck(numeric)
    operator_emoji, operator_name = derive_operator(emoji)
    path = zip_to_path(emoji)
    print(f"deck={deck:02d}, operator={operator_emoji} {operator_name}, path={path}")


def _run_validate(value: str) -> None:
    if validate_zip(value):
        print("VALID")
    else:
        print(f"INVALID ({_invalid_reason(value)})")


def main() -> None:
    parser = argparse.ArgumentParser(description="Convert and validate PPL± zip codes")
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument("--convert", help="Convert an emoji or numeric zip code")
    group.add_argument("--validate", help="Validate an emoji or numeric zip code")
    args = parser.parse_args()

    if args.convert:
        _run_convert(args.convert)
    if args.validate:
        _run_validate(args.validate)


if __name__ == "__main__":
    main()

File: canvas/src/parser/zip-converter.ts
/**
 * Zip Converter — Bidirectional emoji <-> numeric zip code conversion
 *
 * Converts between the 4-emoji SCL zip codes (e.g., '⛽🏛🪡🔵') and their
 * 4-digit numeric equivalents (e.g., '2123'). Also derives deck number
 * and default operator from any zip code.
 *
 * All emoji mappings sourced from CLAUDE.md canonical tables via scl.ts.
 * Zero runtime imports from middle-math/ — zip-registry.json is test-only.
 */

import {
  ORDERS,
  AXES,
  TYPES,
  COLORS,
  ORDER_BY_EMOJI,
  AXIS_BY_EMOJI,
  TYPE_BY_EMOJI,
  COLOR_BY_EMOJI,
  OPERATOR_TABLE,
  PREPARATORY_COLORS,
  EXPRESSIVE_COLORS,
  ORDER_COUNT,
  AXIS_COUNT,
  TYPE_COUNT,
  COLOR_COUNT,
} from '../types/scl';

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/** Digit ranges for each zip position: [label, min, max] */
const RANGES: readonly [string, number, number][] = [
  ['order', 1, ORDER_COUNT],
  ['axis', 1, AXIS_COUNT],
  ['type', 1, TYPE_COUNT],
  ['color', 1, COLOR_COUNT],
];

/**
 * Validates a 4-digit numeric zip string. Throws descriptive errors.
 */
function validateNumericZip(numericZip: string): void {
  if (numericZip.length !== 4 || !/^\d{4}$/.test(numericZip)) {
    throw new Error(`numeric zip must be exactly 4 digits, got '${numericZip}'`);
  }

  for (let i = 0; i < 4; i++) {
    const digit = Number(numericZip[i]);
    const [label, min, max] = RANGES[i]!;
    if (digit < min || digit > max) {
      throw new Error(`${label} ${digit} out of range ${min}-${max}`);
    }
  }
}

/**
 * Detect whether a string looks numeric (all digits).
 */
function isNumeric(s: string): boolean {
  return /^\d+$/.test(s);
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Convert an emoji zip code to its numeric equivalent.
 *
 * @example emojiToZip('⛽🏛🪡🔵') // '2123'
 */
export function emojiToZip(emojiZip: string): string {
  const chars = [...emojiZip];
  if (chars.length !== 4) {
    throw new Error(
      `emoji zip must contain exactly 4 emojis, got ${chars.length}: '${emojiZip}'`
    );
  }

  const order = ORDER_BY_EMOJI[chars[0]!];
  if (order === undefined) throw new Error(`invalid order emoji: ${chars[0]}`);

  const axis = AXIS_BY_EMOJI[chars[1]!];
  if (axis === undefined) throw new Error(`invalid axis emoji: ${chars[1]}`);

  const type = TYPE_BY_EMOJI[chars[2]!];
  if (type === undefined) throw new Error(`invalid type emoji: ${chars[2]}`);

  const color = COLOR_BY_EMOJI[chars[3]!];
  if (color === undefined) throw new Error(`invalid color emoji: ${chars[3]}`);

  return `${order}${axis}${type}${color}`;
}

/**
 * Convert a numeric zip code to its emoji equivalent.
 *
 * @example zipToEmoji('2123') // '⛽🏛🪡🔵'
 */
export function zipToEmoji(numericZip: string): string {
  validateNumericZip(numericZip);

  const order = Number(numericZip[0]) as keyof typeof ORDERS;
  const axis = Number(numericZip[1]) as keyof typeof AXES;
  const type = Number(numericZip[2]) as keyof typeof TYPES;
  const color = Number(numericZip[3]) as keyof typeof COLORS;

  return `${ORDERS[order].emoji}${AXES[axis].emoji}${TYPES[type].emoji}${COLORS[color].emoji}`;
}

/**
 * Check whether a 4-digit numeric zip string addresses a valid zip code.
 *
 * @example isValidZip('2123') // true
 * @example isValidZip('2193') // false (type 9 out of range 1-5)
 */
export function isValidZip(zip: string): boolean {
  try {
    validateNumericZip(zip);
    return true;
  } catch {
    return false;
  }
}

/**
 * Derive the deck number (1-42) from a zip code.
 * Accepts both emoji and numeric input.
 *
 * Formula: (order - 1) * 6 + axis
 *
 * @example zipToDeck('2123') // 7
 * @example zipToDeck('⛽🏛🪡🔵') // 7
 */
export function zipToDeck(zip: string): number {
  const numeric = isNumeric(zip) ? zip : emojiToZip(zip);
  validateNumericZip(numeric);

  const order = Number(numeric[0]);
  const axis = Number(numeric[1]);

  return (order - 1) * AXIS_COUNT + axis;
}

/**
 * Derive the default operator for an emoji zip code.
 * Uses axis (position 2) and color polarity (position 4).
 *
 * @example deriveOperator('⛽🏛🪡🔵') // { emoji: '🤌', name: 'facio' }
 */
export function deriveOperator(emojiZip: string): { emoji: string; name: string } {
  const chars = [...emojiZip];
  if (chars.length !== 4) {
    throw new Error(
      `emoji zip must contain exactly 4 emojis, got ${chars.length}`
    );
  }

  const axisEmoji = chars[1]!;
  const colorEmoji = chars[3]!;

  const axisEntry = OPERATOR_TABLE[axisEmoji];
  if (!axisEntry) {
    throw new Error(`invalid axis emoji for operator derivation: ${axisEmoji}`);
  }

  if (COLOR_BY_EMOJI[colorEmoji] === undefined) {
    throw new Error(`invalid color emoji: ${colorEmoji}`);
  }

  if (PREPARATORY_COLORS.has(colorEmoji)) {
    return { emoji: axisEntry.preparatory.emoji, name: axisEntry.preparatory.latin };
  }

  if (EXPRESSIVE_COLORS.has(colorEmoji)) {
    return { emoji: axisEntry.expressive.emoji, name: axisEntry.expressive.latin };
  }

  throw new Error(`unable to derive polarity for color: ${colorEmoji}`);
}
